File: 18. Implement deque using arrays.py
"""
Task: Deque implemented using circular array
Time Complexity: O(1) operations
Space Complexity: O(k)
"""

class ArrayDeque:
    def __init__(self, k):
        self._data = [None]*k
        self._head = 0
        self._tail = 0
        self._size = 0
        self._cap = k

    def append(self, x):
        if self._size == self._cap:
            return False
        self._data[self._tail] = x
        self._tail = (self._tail+1)%self._cap
        self._size += 1
        return True

    def appendleft(self, x):
        if self._size == self._cap:
            return False
        self._head = (self._head-1)%self._cap
        self._data[self._head] = x
        self._size += 1
        return True

    def pop(self):
        if self._size==0:
            raise IndexError("pop from empty")
        self._tail = (self._tail-1)%self._cap
        val = self._data[self._tail]
        self._data[self._tail]=None
        self._size-=1
        return val

    def popleft(self):
        if self._size==0:
            raise IndexError("popleft from empty")
        val = self._data[self._head]
        self._data[self._head]=None
        self._head = (self._head+1)%self._cap
        self._size-=1
        return val

if __name__=="__main__":
    dq=ArrayDeque(3)
    dq.append(1);dq.append(2);dq.appendleft(0)
    print(dq.pop(), dq.popleft())

# Task: Implement deque using arrays

File: 18. Implement deque using arrays.cpp
/*
 Task: Implement deque using a circular array
*/

#include <bits/stdc++.h>
using namespace std;

class MyDeque
{
    int* arr;
    int cap;
    int fronti;
    int size;

   public:
    MyDeque(int k)
    {
        cap = k;
        arr = new int[cap];
        fronti = 0;
        size = 0;
    }

    ~MyDeque() { delete[] arr; }

    bool push_back(int x)
    {
        if (size == cap)
            return false;
        arr[(fronti + size) % cap] = x;
        size++;
        return true;
    }

    bool push_front(int x)
    {
        if (size == cap)
            return false;
        fronti = (fronti - 1 + cap) % cap;
        arr[fronti] = x;
        size++;
        return true;
    }

    bool pop_back()
    {
        if (size == 0)
            return false;
        size--;
        return true;
    }

    bool pop_front()
    {
        if (size == 0)
            return false;
        fronti = (fronti + 1) % cap;
        size--;
        return true;
    }

    int get_front()
    {
        if (size == 0)
            throw runtime_error("empty");
        return arr[fronti];
    }

    int get_back()
    {
        if (size == 0)
            throw runtime_error("empty");
        return arr[(fronti + size - 1) % cap];
    }

    int get_size() { return size; }
};

int main()
{
    MyDeque dq(5);
    dq.push_back(1);
    dq.push_back(2);
    dq.push_front(0);
    cout << dq.get_front() << " " << dq.get_back() << "\n";
    dq.pop_front();
    cout << dq.get_front() << "\n";
    return 0;
}

// Time Complexity: operations O(1)
// Space Complexity: O(k)
